src/plot_api/plot_api.js

```diff
@@ -195,6 +195,40 @@ Plotly.plot = function(gd, data, layout, config) {
             }
         }
 
+        if(!fullLayout._glcanvas && fullLayout._has('gl')) {
+            fullLayout._glcanvas = fullLayout._glcontainer.selectAll('.gl-canvas').data([{
+                key: 'contextLayer',
+                context: true,
+                pick: false
+            }, {
+                key: 'focusLayer',
+                context: false,
+                pick: false
+            }, {
+                key: 'pickLayer',
+                context: false,
+                pick: true
+            }]);
```

Comment:
I think you need a [key-function](https://bost.ocks.org/mike/constancy/) here to make sure to not create these canvas nodes on every _hard_ redraw that go through `drawFramework`.

Sometime like:

```js
fullLayout._glcanvas = fullLayout._glcontainer.selectAll('.gl-canvas').data([{
  /* .. as currently ... */
}], function(d) { return d.key; });
```

Reply:
This makes me wonder: we probably don't have sufficient test coverage for `Plotly.restyle` calls that morph a gl-based trace types into a SVG-based. For example off the regl2d branch we should have:

```js
Plotly.newPlot(gd, [{
  type: 'scattergl',
  // ...
}])
.then(function() {
   d3.selectAll('canvas').size() // => 3

  return Plotly.restyle(gd, 'type', 'scatter')
})
.then(function() {
  d3.selectAll('canvas').size() // => 0
})
```

As restyling a `parcoords` traces into any other trace types doesn't make much sense at the moment. You could instead try something like:

```js
Plotly.newPlot(gd, [{
  type: 'parcoords',
  // ...
}])
.then(function() {
   d3.selectAll('canvas').size() // => 3

  return Plotly.deleteTraces(gd, [0])
})
.then(function() {
  d3.selectAll('canvas').size() // => 0
})
```